
## Relevant References

### Miscoding, Misloading

* review https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6438450/


### S-AGC -> N mistranslation:

* This may be random and likely known to you; but @Sina Rafiee had
observed some errors in S/N annotation in the file; but we also do
find quite some S->N substitutions in the final set. Among those the
most frequent serine codon in my current AAS codon frequency table was
AGC. This is somewhat consistent with the results reported here for a
very specific case:
https://pubs.acs.org/doi/full/10.1021/ac901541h?casa_token=_XVU6kJg9U8AAAAA%3AIsSLGKZqCFAm13P8lmhc77sxiJHDNKSkkS7d9Hd66Rq2OHZ7vcJh7rNv7tpDnEvbBwDaj8YzGQhsZYs

### S -> BMAA

* https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0075376#s1


## meeting, 20240205

* conservation in Raf,
* AAS in ribosomal proteins,
* evolutionary conservation,
* protein domains.

* UpSet diagram,
* scatter plot, conservation rat:liver?
* cancer vs. normal, variability between cancer.

* large effect sizes that we really believe in.

* ribosome density

# meeting, 20240212

* jaccard distance matrix? SAAP shared between data sets.
* LSCC: standard display set.
* LUAD: higher ratios in tumor than normal throughout.

* proteasome SAAP higher RAAS in all tumors,
  subunit A, SAAP: AGSEELAHQK.

* tissue-type specificity?

* subset of substitutions that only require a wobble
  -> codon vs. AA similarity.

-> group AA-types by similarity, and plot their RAAS distribution


* protein/nucleic acids sequences, 10
  low medium high RAAS from one data set.

-> frequency of each codon in all AAS transcripts,
-> motif search: cut sequences from FASTA
