# Results for Analysis of Amino Acid Substitution Sites by Shiri Tsour et al.

## TODOs:

* Select and adapt 1d and 3d example figures,
* Better version of 2-page GO figure (Ext.Data Fig. 7) from my side?,
* re-do random forest.

## 2024-10-01 - updated files

* `codons/codon_frequencies.tsv`: comparing with the current figure, Fig2j
and Ext.Data Fig. 3l both require update,

* `structure/proteins_raas.tsv `: new mapping to experimental data
accounts for gene name synonyms to increase number of mapped
proteins. Note that this also increases correlation of median protein
RAAS to protein half-lives.

* `aminoacids/AAprop_cancer_dotplot_manual.pdf`: Ext.Data Fig. 5b requires update. 

* `structure/structure_cor_MMSeq2_RAAS.png` and
`structure/structure_cor_MMSeq2_RAAS.png`: Ext. Data Fig. 6f,g require update.

* `function/go_dissection.pdf`: this is a 2-page PDF, Ext.Data Fig. 7 
requires update to 2-page figure.

* `motifs/motifs_cancer_dotplot.pdf`: Fig. 4c requires update,

* `motifs/classes_conservation_disorder_raas.pdf`: Fig 4i requires update,

* `motifs/<motif>.pdf`: Fig. 4d requires update,

* `structure/protein_lengths_all.png`, `structure/protein_halflives_all.png`: Fig 4g,h require update,
or can be generated in paper-style via `structure/proteins_raas.tsv`,

* `sites_raas.tsv`: random forest analysis in Fig. 4j requires update.
