---
fontfamily: helvet
header-includes: 
  - \renewcommand{\familydefault}{\sfdefault}
  - \graphicspath{{/home/raim/data/mistrans/figures/raasprofiles3}}
  - \usepackage[left=2cm,right=2cm,bottom=2cm,top=2cm]{geometry}
  - \pagenumbering{gobble}
bibliography: /home/raim/ref/tata.bib
---

![](motifs/AAS_overlap.pdf){height=35%}![](kraq/peptides_AAS_AAStype_tight.pdf){height=35%}

\vspace{-4ex}

\phantom{das muss noch mehr in die mitteeeeeeeee}![](legend_pvals_horizontal.pdf){width=20%}

\vspace{-50ex} \textbf{a}\phantom{das muss noch mehr in die mitteeeeeeeeeeeeeeee}\textbf{b}

\vspace{46ex}
\textbf{c}

![](motifs/logos/AA_logos_fromto_Q:G_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_Q:A_encoded.pdf){width=19%}
<!-- ![](motifs/logos/AA_logos_fromto_I:G_encoded.pdf){width=19%} prev sig on intron -->
![](motifs/logos/AA_logos_fromto_E:N_encoded.pdf){width=19%}
<!-- ![](motifs/logos/AA_logos_fromto_S:N_encoded.pdf){width=19%} prev sig -->
![](motifs/logos/AA_logos_fromto_T:V_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_N:M_encoded.pdf){width=19%}
<!--  ![](motifs/logos/AA_logos_fromto_T:Q_encoded.pdf){width=19%} prev sig on exon -->

![](motifs/logos/AA_logos_fromto_M:G_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_L:G_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_N:G_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_C:G_encoded.pdf){width=19%}

![](motifs/logos/AA_logos_fromto_E:C_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_I:Q_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_L:Q_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_P:T_encoded.pdf){width=19%}


![](motifs/logos/AA_logos_fromto_E:M_encoded.pdf){width=19%}
<!-- ![](motifs/logos/AA_logos_fromto_M:D_encoded.pdf){width=19%} prev sig -->
![](motifs/logos/AA_logos_fromto_W:D_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_H:D_encoded.pdf){width=19%}
![](motifs/logos/AA_logos_fromto_T:D_encoded.pdf){width=19%}

<!-- \newpage



Extended Data Figure Motifs: \textbf{Amino Acid and Substitution Site
Enrichment Profiles and Sequence Logos.} \textbf{(a)} Counts (text)
and enrichments (gray scale: p-values of cumulative hypergeometric
distribution tests) of amino acids surrounding the amino acid
substitution (AAS) sites. Lysine (K) and Arginine (R), the trypsin
cleavage sites are enriched directly upstream of the substitution
sites. Tryptophane (W), methionine (M), and at lower signficance,
glycine (G) and cytosine (C) are enriched directly adjacent to
substitution sites. \textbf{(b)} As (a) but for substitution types
(encoded:incorporated) vs. their position in identified
peptides. Substitutions by glycine (x:G) or alanine (x:A) are enriched
within the 3 N-temrinal amino acids of base peptides (N1 to N3), i.e.,
directly after the trypsin cleavage sites (K or R). Various
substitutions involving arginine (N), methionine (M) or glutamate (E)
as either the encoded or the incorporated amino acid are enriched
distant from the N- and C-termini (>9). Only substitution types with
at least one significant enrichment ($p\le10^{-10}$) are shown in (b).
\textbf{c.} Sequence difference logos were calculated for all unique
sequences surrounding substitution sites, subset for all observed
substitution types (encoded$\to$incorporated amino acids) using the R
package \texttt{DiffLogo}\citep{[Nettling2015} (version 2.26.0) . Plots were
only generated if any of the positions -3 to +3 around a substitution
site showed a significant enrichment with $p<10^{-10}$, and all these
logos are shown. The logos were grouped by common patterns: (i)
Substitutions by glycine or alanine (Q$\to$A, Q$\to$G, M$\to$G,
L$\to$G) are enriched directly upstream with lysine (K) or arginine
(R), i.e. they are preferentially observed at the N-terminus of base
peptides, next to the trypsin cleavage sites (K or R). (ii)
Substitutions of glutamine (Q$\to$A) or ariginine (N$\to$G) are
flanked by cysteine (C) enrichments. (iii) substitutions E$\to$N,
T$\to$V and N$\to$M are flanked by methionine (M) enrichments. (iv)
Substitutions EC, IQ, LQ and PT are flanked by tryptophane (W)
enrichments. To generalize these grouped observations, we defined four
sequence classes used for the difference logos in Figure 4e, where
sequence difference logos (each against all other sequences in our
set) were calculated for all substitutions by G or A within the 3
N-terminal sites of base peptides (motif \texttt{KRAQ}), and for all
substitutions that had at least one W (\texttt{WWxWW}), M
(\texttt{MMxMM}), C (\texttt{CCxCC}) or G (\texttt{GGxGG}) within 2
positions of the substitution site. The selection \texttt{GGxGG}
showed no specific enrichments at the subsitution site and is not
shown.

-->

<!-- These observations (a-c) lead to the definition of sequences for Figure 4e,
where sequence difference logos were calculated for all substitutions
by G or A within the 3 N-terminal sites of base peptides (motif
\texttt{KRAQ}), and for all substitutions that had at least one W
(\texttt{WWxWW}), M (\texttt{MMxMM}), C (\texttt{CCxCC}) or G
(\texttt{GGxGG}) within 2 positions of the substitution site. The
selection \texttt{GGxGG} showed no specific enrichments at the
subsitution site and is not shown.} -->
