
require(segmenTools)
require(qvalue)
library(viridis)

### PLOT UTILS

## COLOR FUNCTION ARNO
## generate colors similar to inferno but with
## a better yellow (viridis)
mcol <- viridis::inferno(5)
vcol <- viridis::viridis(5)
mcol[5] <- vcol[5]
arno <- colorRampPalette(mcol)

## NOTE: adapted from SpacePAC,
## * multiple spheres
plotRaas3d <- function (position.matrix, center, radius,
                        color, name = "", alpha = 1) 
{
    require(rgl)
    .check3d()
    ##open3d()
    bg3d(color = c("white"))
    plot3d(x = position.matrix[, 4], y = position.matrix[, 5], 
        z = position.matrix[, 6], type = "s", size = 0.5, add = FALSE, 
        xlab = "", ylab = "", zlab = "")
    plot3d(x = position.matrix[, 4], y = position.matrix[, 5], 
        z = position.matrix[, 6], type = "l", add = TRUE, col = "blue")
    decorate3d(main = name, axes = TRUE, xlab = "x-axis", 
               ylab = "y-axis", zlab = "z-axis")

    if ( length(radius)==1 )
        radius <- rep(radius, length(center))
    if ( missing(color) ) color <- "red"
     if ( length(color)==1 )
        color <- rep(color, length(center))
   
    for ( c in seq_along(center) ) {
        index <- which(position.matrix[, 2] == center[c])

        spheres3d(x = position.matrix[index, 4],
                  y = position.matrix[index, 5],
                  z = position.matrix[index, 6],
                  radius = radius[c], color = color[c], 
                  alpha = alpha)
    }
}

# from->to as expression, e.g. for axis labels
ftlabels <- function(srt) {
    axex <- rep("",length(srt))
    names(axex) <- srt
    for ( i in seq_along(srt) ) {
        ft <- unlist(strsplit(srt[i],":"))
        axex[i] <- as.expression(bquote(.(ft[1]) %->% .(ft[2])))
    }
    axex
}

## currently unused
addPoints <- function(ovl, value="median") {
    z <- t(apply(ovl[[value]], 2, rev))
    cx <- c(z)
    mx <- max(cx,na.rm=TRUE)
    mn <- min(cx,na.rm=TRUE)
    cx <- 3*(cx-mn)/(mx-mn)
    points(x = rep(1:ncol(z), nrow(z)),
           y = rep(nrow(z):1, each = ncol(z)), cex = cx)
}

### STATISTICS 

## wilcox test with normalized U-statistic, to
## be used for statistical profile plots
w.test <- function(x,y) {
    res <- wilcox.test(x,y)
    ## normalized U-statistic
    tt <- res$statistic/(sum(!is.na(x))*sum(!is.na(y))) -0.5
    rt <- list()
    rt$statistic <- unlist(tt)
    rt$p.value <- unlist(res$p.value)
    rt
}


### HIGH LEVEL PLOT FUNCTIONS

dotprofile <- function(x, value, vcols=viridis::viridis(100),
                       vbrks, p.dot=1e-10, dot.sze=c(.3,2), xpd=FALSE,
                       lg2=FALSE, mxr,
                       show.total=FALSE, tot.cex=.8,
                       test=FALSE, ...) {

    ## values for coloring
    vals <- x[[value]]
    if ( lg2 ) {
        vals <- log2(vals)
        if ( missing(mxr) ) mxr <- max(abs(vals))
        vals[vals >  mxr] <-  mxr
        vals[vals < -mxr] <- -mxr
    }

    ## breaks
    if ( missing(vbrks) )
        vbrks <- seq(min(vals,na.rm=TRUE), max(vals, na.rm=TRUE),
                     length.out=length(vcols)+1)

    navals <- vals
    navals[] <- NA
    image_matrix(navals, breaks=vbrks, col=vcols, ...)

    p <- x$p.value
    p <- -log10(p)
    p[p>-log10(p.dot)] <- -log10(p.dot)
    z <- p/-log10(p.dot)


    
    ## intersect data to colors
    cols <- vcols[findInterval(t(vals),
                               seq(min(vbrks), max(vbrks),
                                   length.out=length(vbrks)),
                               all.inside = TRUE)]
    d.sze <- dot.sze[1]+dot.sze[2]*c(t(z))
    if ( test )
        points(x = rep(1:ncol(z), nrow(z)),
               y = rep(nrow(z):1, each= ncol(z)), cex=d.sze)
    points(x = rep(1:ncol(z), nrow(z)),
           y = rep(nrow(z):1, each= ncol(z)),
           cex=d.sze, pch=19,
           col=cols, xpd=xpd)
    
    toty <- totx <- FALSE
    if (is.logical(show.total)) {
        if (show.total) 
            toty <- totx <- TRUE
    }
    else if (is.character(show.total)) {
        if (show.total == "x") 
            totx <- TRUE
        if (show.total == "y") 
            toty <- TRUE
        if (show.total %in% c("xy", "yx")) 
            toty <- totx <- TRUE
    }
    if (toty) 
        if ("num.query" %in% names(x)) 
            axis(4, at = length(x$num.query):1,
                 labels = format(x$num.query, big.mark=",", trim=TRUE), 
                las = 2, lwd = 0, lwd.ticks = 1, cex.axis=tot.cex)
    if (totx) 
        if ("num.target" %in% names(x)) 
            axis(3, at = 1:length(x$num.target),
                 labels = format(x$num.target, big.mark=",", trim=TRUE), 
                 las = 2, lwd = 0, lwd.ticks = 1, cex.axis=tot.cex)
    ##if (toty | totx) 
    ##    figlabel("total", region = "figure", pos = "topright", 
    ##        cex = par("cex"))
}

## testing to use enrichment of overlap tables
## as dot size -> too big for low numbers
plotOverlaps2 <- function(ovw,
                          col,
                          max.val=3, 
                          p.min=1e-10, p.txt=1e-5,
                          dot.sze=c(.2,3), bg=TRUE,
                          ...) {
    
    ## p-value scaling
    p <- -log10(ovw$p.value)
    p[p>-log10(p.min)] <- -log10(p.min)
    p <- p/-log10(p.min)

    p <- p*ovw$sign
    
    ## enrichment ratio scaling
    r <- abs(log2(ovw$ratio))
    r[r>max.val] <- max.val
    r <- r/max.val
    

    ## point size
    d.cex <- dot.sze[1]+dot.sze[2]*c(t(r))


    ## p-value colors
    if ( missing(col) ) {
        docols <- colorRampPalette(c("#FFFFFF","#0000FF"))(50)
        upcols <- colorRampPalette(c("#FFFFFF","#FF0000"))(50)
        col <- unique(c(rev(docols), upcols))
    }
    brks <- seq(-1, 1, length.out = length(col)+1)
    cols <- col[findInterval(t(p), brks, all.inside = TRUE)]

    navals <- ovw[["ratio"]]
    navals[] <- NA
    image_matrix(navals,  col=col, breaks=brks, ...)

    if ( bg )
        points(x = rep(1:ncol(r), nrow(r)),
               y = rep(nrow(r):1, each= ncol(r)), cex=d.cex+.5, pch=1)

    points(x = rep(1:ncol(r), nrow(r)),
           y = rep(nrow(r):1, each= ncol(r)),
           cex=d.cex, pch=19,
           col=cols)
    return(cbind(c(p),cols))
}

plotProfiles <- function(ovw, mai=c(.6,.5,.5,.5),
                         fw=.2, fh=.2,
                         ttcols=ttcols, p.min=p.min, p.txt=p.txt,
                         dot.sze=dot.sze, p.dot=p.dot, ## TODO: legend
                         value="median", vcols, vbrks,
                         count="unique", gcols=gcols,
                         fname="profile",
                         mtxt, mtxt1, mtxt.line=1.3, mtxt.cex=1,
                         llab, rlab, xlab=expression(log[10](RAAS)),
                         ffam="sans",
                         ftyp="png",
                         axis1.col,
                         axis2.col,
                         axis1.las=2,
                         bg="white",
                         tot.cex=.8,
                         col.lines, # column classes - vertical lines
                         plot.all=FALSE, plot.legend=FALSE, verb=0) {
    
    ## replace row labels with arrows
    ## TODO: adapt to plot by AA colors
    rows <- rownames(ovw$p.value)
    axex <- rep("",length(rows))
    names(axex) <- rows
    for ( i in seq_along(rows) ) {
        if ( length(grep(":",rows[i])) ) {
            ft <- unlist(strsplit(rows[i],":"))
            ft[ft=="x"] <- ""
            axex[i] <- as.expression(bquote(.(ft[1]) %->% .(ft[2])))
        } else axex[i] <- as.expression(bquote(.(rows[i])))
    }
      
    ## calculate optimal figure height: result fields + figure margins (mai)
    nh <- nrow(ovw$p.value) *fh + mai[1] + mai[3]
    nw <- ncol(ovw$p.value) *fw + mai[2] + mai[4]

    ## p-value profiles

    if ( missing(vbrks) ) {
        vals <- ovw[[value]]
        vals <- vals[!is.na(vals)]
        vbrks <- seq(min(vals), max(vals), length.out=101)
        vcols <- viridis::viridis(100)
    }

    ## combined effect size and p-value plot
    if ( verb ) cat(paste("plotting dotplot\n"))
    plotdev(paste0(fname,"_dotplot"),
            height=nh, width=nw, res=300, type=ftyp, bg=bg)
    par(mai=mai, mgp=c(1.3,.3,0), tcl=-.25, family=ffam)
    dotprofile(x=ovw, value=value, vbrks=vbrks,
               vcols=vcols, p.dot=p.dot,
               dot.sze=dot.sze,
               axis=NA, xlab=NA, ylab=NA)
    box()
    if ( !missing(axis1.col) )
        Map(axis, 1, at=1:ncol(ovw$p.value),
            labels=colnames(ovw$p.value), las=axis1.las, family=ffam,
            col.axis=axis1.col, font=2)#, cex.axis=1.2)
    else
        axis(1,  1:ncol(ovw$p.value),
             labels=colnames(ovw$p.value), las=axis1.las, family=ffam)
     if ( !missing(axis2.col) ) {
         Map(axis, 2, length(axex):1, labels=axex, las=2, family=ffam,
             col.axis=axis2.col[rows], font=2, cex.axis=1.2)
     }
    else
        axis(2, length(axex):1, labels=axex, las=2, family=ffam)
    axis(3, at=1:ncol(ovw$num.target), cex.axis=tot.cex,
         labels=format(ovw$num.target[1,], big.mark=",", trim=TRUE),las=2)
    axis(4, at=nrow(ovw$num.query):1, cex.axis=tot.cex,
         labels=format(ovw$num.query[,1], big.mark=",", trim=TRUE),las=2)
    if ( !missing(mtxt) ) mtext(mtxt, 2, mtxt.line, cex=mtxt.cex)
    if ( !missing(mtxt1) ) mtext(mtxt1, 1, mtxt.line, cex=mtxt.cex)
    if ( !missing(llab) ) figlabel(llab, pos="bottomleft", cex=1.2)
    if ( !missing(rlab) ) figlabel(rlab, pos="bottomright", cex=.8)
    if ( !missing(col.lines) )
        abline(v=.5+col.lines, lwd=1, xpd=FALSE)
    dev.off()
    ##return(p)


    ## plot size adjusted legend
    if ( plot.legend ) {

        ## vcols, vbreaks, p.min, p.dot, dot.sze
        ## tigtht RAAS range - legend for acols/abreaks
        pp <- seq(0, -log10(p.dot), length.out=3)
        rs <- unique(round(seq(min(vbrks),max(vbrks),length.out=4) ))
        pm <- matrix(rep(pp, each=length(rs)), nrow=length(rs))
        rm <- matrix(rep(rs, length(pp)), ncol=length(pp))
        colnames(pm) <- colnames(rm) <- -pp
        rownames(pm) <- rownames(rm) <- round(rs,1)
        ovlg <- list(p.value=t(10^-pm), median=t(rm))
        
        lmai <- c(.4,.5,.05,.06)
        fh <- fw <- .2
        lh <- nrow(ovlg$p.value) *fh + lmai[1] + lmai[3]
        lw <- ncol(ovlg$p.value) *fw + lmai[2] + lmai[4]
        
        plotdev(paste0(fname,"_dotplot_legend"),
                height=lh, width=lw, res=300, type=ftyp, bg=bg)
        par(mai=lmai, mgp=c(1.3,.3,0), tcl=-.25)
        dotprofile(ovlg, value="median",
                   vbrks=vbrks,
                   vcols=vcols, 
                   dot.sze=dot.sze, p.dot=p.dot, axis=1:2,
                   ylab=plab,
                   xlab=NA)
        ##mtext(xl.raas, 1, 1.1, adj=-.4)
        text(1.5, -1, xlab, xpd=TRUE)
        dev.off()
    }
    
    if ( !plot.all ) return()
    
    if (verb>0) cat(paste("plotting",value,"counts\n"))
    ## BARPLOT OF COUNTS
    mai.bar <- mai
    ##mai.bar[c(2,4)] <- mai.bar[c(2,4)] +.05
    mai.bar[c(1,3)] <- .075
    plotdev(paste0(fname,"_ccounts"),
            height=.75, width=nw, res=300, type=ftyp)
    par(mai=mai.bar, mgp=c(1.3,.3,0), tcl=-.25, xaxs="i", family=ffam)
    bp <- barplot(c(ovw$num.target), axes=FALSE, xlab=NA, #beside=TRUE,
                  ylab="", las=2, xaxt='n', 
                  width=.5, space=.5)
    ##mtext(expression(count), 2, 2.5)
    if ( !missing(col.lines) ) {
        cdn.cn <- head(col.lines, length(col.lines)-1)
        abline(v=bp[cdn.cn,]+unique(diff(bp[,1]))/2)
    }
    for ( ax in c(2,4) ) 
        axis(ax, las=2)
    dev.off()
    mai.bar <- mai
    mai.bar[c(2,4)] <- .075
    ##mai.bar[c(1,3)] <- .075
    plotdev(paste0(fname,"_rcounts"),
            height=nh, width=.75, res=300, type=ftyp)
    par(mai=mai.bar, mgp=c(1.3,.3,0), tcl=-.25, yaxs="i", family=ffam)
    bp <- barplot(rev(c(ovw$num.query)), axes=FALSE, xlab=NA,
                  horiz=TRUE,
                  ylab="", las=2, yaxt='n', 
                  width=.5, space=.5)
    ##mtext(expression(count), 2, 2.5)
    for ( ax in c(1,3) )
        axis(ax, las=2)
    dev.off()

    ## classical p value profile
    outf <- paste0(fname,"_wtests")
    if (verb>0) cat(paste("plotting test profile",outf,p.min,p.txt,"\n"))
    plotdev(outf, height=nh, width=nw, res=300, type=ftyp)
    par(mai=mai, mgp=c(1.3,.3,0), tcl=-.25)
    plotOverlaps(ovw, p.min=p.min, p.txt=p.txt,
                 text.cex=.8, axis=NA, ylab=NA, xlab=NA,
                 col=ttcols, show.total=TRUE)
    if ( !missing(axis2.col) )
        Map(axis, 2, length(axex):1, labels=axex, las=2, family=ffam,
            col.axis=axis2.col[rows])
    else
        axis(2, length(axex):1, labels=axex, las=2, family=ffam)
    if ( !missing(axis1.col) )
        Map(axis, 1, 1:ncol(ovw$p.value),
            labels=colnames(ovw$p.value), las=2, family=ffam,
            col.axis=axis1.col)
    else
        axis(1,  1:ncol(ovw$p.value),
            labels=colnames(ovw$p.value), las=2, family=ffam)
    if ( !missing(mtxt) ) mtext(mtxt, 2, mtxt.line, cex=mtxt.cex)
    if ( !missing(mtxt1) ) mtext(mtxt1, 1, mtxt.line, cex=mtxt.cex)
    if ( !missing(llab) ) figlabel(llab, pos="bottomleft", cex=1.2)
    if ( !missing(rlab) ) figlabel(rlab, pos="bottomright", cex=.8)
    if ( !missing(col.lines) )
        abline(v=.5+col.lines, lwd=1, xpd=FALSE)
    dev.off()
    
    if (verb>0) cat(paste("plotting",value,"RAAS\n"))
    
    plotdev(paste0(fname,"_raas_",value),
            height=nh, width=nw, res=300, type=ftyp)
    par(mai=mai, mgp=c(1.3,.3,0), tcl=-.25)
    txt <- ovw[["count"]]
    txt[txt==0] <- ""
    q1 <- quantile(c(ovw[[value]]), probs=.4, na.rm=TRUE)
    txt.col <- ifelse(ovw[[value]]<q1, "white","black")
    image_matrix(ovw[[value]], col=vcols, breaks=vbrks, axis=NA,
                 text=txt, text.col=txt.col, xlab=NA, ylab=NA, text.cex=.8)
    if ( !missing(axis1.col) )
        Map(axis, 1, 1:ncol(ovw$p.value),
            labels=colnames(ovw$p.value), las=2, family=ffam,
            col.axis=axis1.col)
    else
        axis(1,  1:ncol(ovw$p.value),
            labels=colnames(ovw$p.value), las=2, family=ffam)
     if ( !missing(axis2.col) )
        Map(axis, 2, length(axex):1, labels=axex, las=2, family=ffam,
            col.axis=axis2.col[rows])
    else
        axis(2, length(axex):1, labels=axex, las=2, family=ffam)
    if ( !missing(mtxt) ) mtext(mtxt, 2, mtxt.line, cex=mtxt.cex)
    if ( !missing(mtxt1) ) mtext(mtxt1, 1, mtxt.line, cex=mtxt.cex)
    if ( !missing(llab) ) figlabel(llab, pos="bottomleft", cex=1.2)
    if ( !missing(rlab) ) figlabel(rlab, pos="bottomright", cex=.8)
    if ( !missing(col.lines) )
        abline(v=.5+col.lines, lwd=1, xpd=FALSE, col="white")
    dev.off()

    if (verb>0) cat(paste("plotting unique number\n"))
   
    plotdev(paste0(fname,"_count_unique"),
                         height=nh, width=nw, res=300, type=ftyp)
    par(mai=mai, mgp=c(1.3,.3,0), tcl=-.25)
    cnt <- ovw$unique
    cnt[cnt==0] <- NA
    txt <- ovw$unique
    txt[txt==0] <- ""
    txt.col <- ifelse(ovw$unique>quantile(c(ovw$unique),.95),"white","black")
    image_matrix(cnt, col=gcols, axis=1:2,
                 text=txt, text.col=txt.col, ylab=NA, xlab=NA, text.cex=.8)
    ##axis(2, length(axex):1, labels=axex, las=2)
    if ( !missing(mtxt) ) mtext(mtxt, 2, mtxt.line, cex=mtxt.cex)
    if ( !missing(mtxt1) ) mtext(mtxt1, 1, mtxt.line, cex=mtxt.cex)
    if ( !missing(llab) ) figlabel(llab, pos="bottomleft", cex=1.2)
    if ( !missing(rlab) ) figlabel(rlab, pos="bottomright", cex=.8)
    if ( !missing(col.lines) )
        abline(v=.5+col.lines,
               lwd=1, xpd=FALSE)
    dev.off()

    if (verb>0) cat(paste("plotting volcano\n"))

    plotdev(paste0(fname,"_volcano"),
                         height=3, width=4, res=300, type=ftyp)
    par(mai=c(.5,.75,.1,.75), mgp=c(1.3,.3,0), tcl=-.25)
    volcano(ovw, cut=100, p.txt=-log10(p.txt),
            v.txt=c(-Inf,-1), density=TRUE,
            xlab=paste(value,"TMT RAAS"), value=value)
    abline(v=get(value, mode="function")(ovw[[value]],na.rm=TRUE))
    if ( !missing(llab) ) figlabel(llab, pos="bottomleft", cex=1.2)
    if ( !missing(rlab) ) figlabel(rlab, pos="bottomright", cex=.8)
    dev.off()
    
    if (verb>0) cat(paste("plotting histograms\n"))

    ## common histogram with color by sign and
    ## first, calculate densities
    dns <- ovw$ALL
    for ( i in 1:nrow(ovw$p.value) ) {
        for ( j in 1:ncol(ovw$p.value) ) {
            rw <- rownames(ovw$p.value)[i]
            cl <- colnames(ovw$p.value)[j]
            vls <- ovw$ALL[[rw]][[cl]]
            if (length(vls)>1 ) {
                dns[[rw]][[cl]] <- density(vls)
            }
        }
    }
    ## get max
    mxs <- rep(NA, length(dns))
    for ( i in seq_along(dns) )
        mxs[i] <- max(unlist(lapply(dns[[i]],
                                    function(z)
                                        ifelse("y"%in%names(z),
                                               max(unlist(z["y"])),-10))))
    
    plotdev(paste0(fname,"_densities"),
                         res=300, width=4, height=2, type=ftyp)
    par(mai=c(.5,.5,.1,.1), mgp=c(1.3,.3,0), tcl=-.25)
    dns <- ovw$ALL
    hist(unlist(dns), col="#77777755", border=NA, 
         xlim=c(-6,4),
         xlab=expression(TMT~level~log[10]*RAAS), ylab=NA, main=NA, axes=FALSE)
    par(new=TRUE)
    plot(1, col=NA, xlim=c(-6,4), ylim=c(0,max(mxs)),
         xlab=NA, ylab=NA, axes=FALSE)
    for ( i in 1:nrow(ovw$p.value) ) {
        for ( j in 1:ncol(ovw$p.value) ) {
            rw <- rownames(ovw$p.value)[i]
            cl <- colnames(ovw$p.value)[j]
            vls <- ovw$ALL[[rw]][[cl]]
            if (length(vls)>1 ) {
                dns[[rw]][[cl]] <- density(vls)
                if ( ovw$p.value[i,j] <1e-3 )
                    lines(dns[[rw]][[cl]],
                          col=ifelse(ovw$sign[i,j]==-1,"blue","red"),
                          lwd=-log10(ovw$p.value[i,j])/-log10(p.min))
                ## TODO: annotate
            }
        }
    }
    axis(2)#, labels=NA)
    mtext("density",2,1.3)
    axis(1)
    dev.off()
}

## plot p.values, RAAS, and count.
plotovl <- function(ovl, text="count", cut=TRUE, value="median",
                    txt.cut=-2, ...) {

    ## TODO: auto-select txt.cut, in image_matrix based on brightness,
    ## as in 
    txt <- ovl[[text]]
    txt[txt==0] <- ""
    txt.col <- ifelse(ovl[[value]]< txt.cut, "white","black")
    image_matrix(ovl[[value]], cut=cut,
                 text=txt, text.col=txt.col, ...)
}

## volcano plot function for overlap class
## produced by raasProfile
volcano <- function(ovl, cut=15, value="median", pid="p.value",
                    p.txt=6, v.txt, mid,
                    ids, lg2=FALSE, mxr, density=TRUE, xlab, ...) {

    ## values for coloring
    if ( inherits(ovl, "clusterOverlaps") ) { # for clusterOverlaps class
        
        tmv <- gdata::unmatrix(t(ovl[[value]]))
        tpv <-  gdata::unmatrix(t(ovl[[pid]]))

        ## clean plot names
        if ( ncol(ovl[[pid]])==1 )
            names(tpv)<- sub(".*:", "", names(tpv))
        if ( nrow(ovl[[pid]])==1 )
            names(tpv)<- sub(":.*", "", names(tpv))
 
    } else if ( inherits(ovl, "matrix") | inherits(ovl, "data.frame") ) {
        tmv <- ovl[,value]
        tpv <- ovl[,pid]
        ## used for name plotting
        names(tpv) <- rownames(ovl)
    }

    ## process x values
    if ( lg2 ) {
        tmv <- log2(tmv)
        if ( missing(mxr) ) mxr <- max(abs(tmv))
        tmv[tmv >  mxr] <-  mxr
        tmv[tmv < -mxr] <- -mxr
    }
    ## process p values
    tpv <- -log10(tpv)
    tpv[tpv>cut] <- cut

    na <- is.na(tmv) | is.na(tpv)
    tpv <- tpv[!na]
    tmv <- tmv[!na]

    if ( length(tmv)<2 ) 
        density <- FALSE

    if ( missing(xlab) ) xlab <- value
    
    if ( density ) {
        dense2d(tmv, tpv, ylab=expression(-log[10](p)), xlab=xlab, ...)
    } else {
        plot(tmv, tpv, ylab=expression(-log[10](p)), xlab=xlab, ...)
    }
    axis(4)

    ## indicate high
    show <- tpv>p.txt
    if ( !missing(v.txt) )
        show <- show & (tmv<v.txt[1] | tmv>v.txt[2])

    show <- which(show)
    if ( length(show) ) {

        ## optional plot IDs
        labels <- names(tpv)[show]
        if ( !missing(ids) )
            labels <- ids[labels]
        ## vertical line
        if ( missing(mid) )
            mid <- mean(tmv,na.rm=TRUE)
        ## plot labels
        segmenTools::shadowtext(tmv[show], tpv[show], labels=labels,
                                pos=ifelse(tmv[show]<mid, 2,4), col="red",
                                font=2, cex=.8, xpd=TRUE)
    }
    invisible(names(tpv)[show])
}

## calculate statistical profiles, similar to segmenTools::clusterProfiler,
## but working on lists of unequal lengths instead of a matrix
raasProfile.old <- function(x=cdat, id="SAAP", values=tmt,
                        bg=FALSE, vid,
                        rows="to", cols="aacodon",
                        row.srt, col.srt,
                        use.test=use.test,
                        do.plots=interactive(),
                        xlab="value", fname="profile_",
                        verb=FALSE) {

    ## TODO: use full genetic code for consistent columsn and rows!

    if ( !missing(row.srt) ) aas <- row.srt
    else aas <- sort(unique(x[,rows]))
    if ( !missing(col.srt) ) acod <- col.srt
    else acod <- sort(unique(x[,cols]))

    ## FILTER SRTS
    aas <- aas[aas%in%x[,rows]]
    acod <- acod[acod%in%x[,cols]]
    
    codt <- list()
    
    tt <- matrix(0, ncol=length(acod), nrow=length(aas))
    colnames(tt) <- acod
    rownames(tt) <- aas
    tc <- tp <- tm <- td <- tt
    tp[] <- 1
    tm[] <- td[] <- NA


    min.obs=2
    allvals <- list()
    for ( i in seq_along(aas) ) {

        aa <- aas[i]
        allvals[[aa]] <- list()

        for ( j in seq_along(acod) ) {

            cod <- acod[j]
        
            ## get all SAAP for this codon
            idx <- x[,cols]==cod & x[,rows]==aa

            ## NOTE: unique SAAP!!
            csap <- x[idx,id]
            csap <- unique(csap)

            if ( length(idx)==0 ) next
            if ( sum(idx,na.rm=TRUE)==0 ) next
            
            if (verb>0) cat(paste("testing",cod,"to", aa, "\n"))

            ## if a local background is chosen, tmt
            ## should be a table with the bg column
            ## and we filtering TMT only for this subset
            if ( bg ) {

                ## TODO: CHECK THIS: split by SAAP, get all RAAS
                ## duplicate SAAP effect?
                if ( verb>0 )
                    cat(paste("getting subset of values:",
                              cols, cod, vid, id,"\n"))
                vals <- values[values[,cols]==cod,] ## FILTER DATASET
                vals <- split(vals[,vid], vals[,id])## split RAAS/vid by SAAP/id
            } else vals <- values
                
            
            ## get all RAAS values
            rm <- !csap%in%names(vals)
            if ( sum(rm) ) 
                cat(paste("WARNING:", aa, cod, ":", sum(rm), "of", length(csap),
                          "IDs not found in values:",
                          paste(csap[rm],collapse=","),"\n"))
            csap <- csap[!rm]
            
            ## get all RAAS values
            y <- unlist(vals[csap])
            X <- unlist(vals)

            allvals[[aa]][[cod]] <- y
            
            tc[i,j] <- length(y)

            if ( length(y)==0 ) next
            
            tm[i,j] <- mean(y)
            td[i,j] <- median(y)
           
            if ( sum(!is.na(y)) >= min.obs ) {
                tts <- use.test(y, X)
                codt[[cod]] <- tts
                tt[i,j] <- tts$statistic
                tp[i,j] <- tts$p.value
                
                if ( do.plots ) {
                    outfile <- paste0(fname,cod,"_",aa,".png")
                    hcol <- ifelse(tts$statistic<0, "#0000ff","#ff0000")
                    if ( verb>0 ) cat(paste("plotting", outfile, "\n"))
                    png(outfile, width=3, height=2,
                        units="in", res=100)
                    par(mai=c(.5,.15,.25,.15), tcl=-.25, mgp=c(1.3,.3,0))
                    brks <- seq(min(X), max(X), length.out=20)
                    hist(y, freq=FALSE, border=NA,
                         col=paste0(hcol,77),
                         xlim=range(X), main=NA,
                         xlab=xlab, breaks=brks, axes=FALSE)
                    par(new=TRUE)
                    hist(X, freq=FALSE, col="#77777777", border=NA,
                         xlim=range(X), main=NA, xlab=NA, ylab=NA, axes=FALSE,
                         breaks=brks)
                    abline(v=median(X), col=1, lwd=2)
                    abline(v=td[i,j], col=hcol, lwd=2)
                    axis(1)
                    mtext(as.expression(bquote(.(cod) %->% .(aa))),
                          3,-.1, font=2, cex=1.5)
                    legend(ifelse(tts$statistic<0, "topright","topleft"),
                           c(paste0("n=", length(y)),
                             paste0("u/t=", round(tts$statistic,1)),
                             paste0("p=", signif(tts$p.value,1))),
                           bty="n", seg.len=0, x.intersp=0)
                    dev.off()
                }
            }
        }
    }
    if ( verb>0 ) cat(paste("DONE\n"))
    
    ## construct overlap object
    ova <- list()
    ova$p.value <- tp
    ova$statistic <- tt
    ova$count <- tc
    ova$mean <- tm
    ova$median <- td

    ## unique SAAP from main data table
    ova$unique <- table(x[,rows], x[,cols])[rownames(tp),colnames(tp)]
    
    sg <- sign(tt)
    sg[is.na(sg)] <- 1
    sg[sg==0] <- 1
    ova$sign <- sg
    
    ##ova$statistic[is.na(ova$statistic)] <- ""
    
    ## TODO: add counts
    ova$num.target <-
        t(as.matrix(table(x[,cols])[colnames(tp)]))
    ova$num.query <-
        as.matrix(table(x[,rows]))[rownames(tp),,drop=FALSE]


    ## attach all raw values
    ova$ALL <- allvals
    
    class(ova) <- "clusterOverlaps"
    ova
}

## NOTE: CV for log-normal data, see @Canchola2017
## cv = sqrt(10^(log(100)*var) -1)
cvl <- function(x) sqrt(10^(log(10)*var(x,na.rm=TRUE)) -1)

## mean, median, and unpaired t/w tests
listProfile <- function(x, y, delog=TRUE, use.test=t.test, min=3) {

    stat <- lapply(x, function(x) {
        tt=NA
        tp=NA
        if ( length(x)>=min) {
            tts <- use.test(x, y)
            tt <- unname(tts$statistic)
            tp <- tts$p.value
        }
        lx <- x
        if ( delog ) 
            lx <- 10^x
        xmn <- mean(lx)
        xsd <- sd(lx)
        xcv <- xsd/xmn
        xcvl <- cvl(lx)
        xmd <- median(lx)
        if ( delog ) {
            xmn <- log10(xmn)
            xmd <- log10(xmd)
            xsd <- log10(xsd)
        }
        c(mean=xmn, median=xmd,
          sd=xsd, cv=xcv, cvl=xcvl,
          statistic=tt, p.value=tp,
          n=length(x))
    })
    as.data.frame(do.call(rbind, stat))
}

## merge clusterOverlap objectes from raasProfile
## TODO: generalize and move to segmenTools
mergeProfiles <- function(ovll) {

    ## result list that will contain all merged
    ## entries
    ovl <- list()

    ## content of invididual clusterOverlap structures
    flags <- c("delog", "p.adjust")
    mats <- c("p.value", "statistic", "count",
              "mean", "median", "unique", "sign", "num.query")
    lsts <- c("ALL")

    ## collect flags and test for concistency
    for ( flag in flags ) {
        ovl[[flag]] <- unlist(lapply(ovll, function(y) y[[flag]]))
        ## check consistency of flags
        if ( length(unique(ovl[[flag]]))>1 )
            stop("inconsistent processing flags")
        ovl[[flag]] <- unique(ovl[[flag]])
    }
    
    ## collect numbers of targets (columns)
    ## NOTE: only for SAME TARGETS/COLUMNS
    ## TODO: check whether target or query is consistent and fuse accordingly
    num.target <- do.call(rbind, lapply(ovll, function(x) x$num.target))
    if ( any(apply(num.target, 2, sd)>0) )
        stop("inconsistent number of targets (columns)")
    ovl$num.target <- t(as.matrix(apply(num.target, 2, unique)))

    ## rbind all matrices
    for ( mat in mats ) 
        ovl[[mat]] <- do.call(rbind, lapply(ovll, function(x) x[[mat]]))

    ## append lists
    for ( lst in lsts ) {
        tmp <- list()
        for ( i in 1:length(ovll) )
            tmp <- append(tmp, ovll[[i]][[lst]])
        ovl[[lst]] <- tmp
    }
    class(ovl) <- "clusterOverlaps"
    ovl
}

raasProfile <- function(x, rows, ...) {
    
    ## catch traditional case for nonoverlapping class in
    ## a single row.
    if ( inherits(rows, "character") )
        return(raasProfile.row(x=x, rows=rows, ...))

    cat(paste("NOTE: using loop over matrix of classes\n"))

    ## loop through rows matrix
    covw <- list()
    for ( j in 1:ncol(rows) ) {
                
        mid <- colnames(rows)[j]
        mclass <- rep("n.a.", nrow(rows))
        mclass[rows[,j]] <- mid
        
        x$TEST <- mclass

        ovw <- raasProfile(x=x, 
                           rows="TEST", row.srt=mid, ...)
        covw[[mid]] <- ovw
    }
    mergeProfiles(covw)
}

## calculate statistical profiles, similar to segmenTools::clusterProfiler,
## but working on lists of unequal lengths instead of a matrix
raasProfile.row <- function(x=tmtf, id="SAAP", 
                            value="RAAS", delog=TRUE, replace=TRUE,
                            bg=FALSE, bg.dir="col", na.rm=FALSE,
                            rows="to", cols="aacodon",
                            row.srt, col.srt, filter=TRUE,
                            use.test=use.test, p.adjust="none",
                            min.obs=2,
                            do.plots=FALSE,
                            xlab="value", fname="profile_",
                            verb=FALSE) {

    if ( missing(cols) ) {
        x$call <- "all"
        cols <- "call"        
    }
    if ( missing(cols) ) {
        x$rall <- "all"
        cols <- "rall"        
    }

    ## check presence
    if ( any(!c(value, rows, cols)%in%colnames(x)) )
        stop("one of the requested values ",
             paste(value, rows, cols, sep=";"),
             " is not present in the data",
             paste(colnames(x), collapse=";"))
    
    ## sorting of row and column classes
    if ( !missing(row.srt) ) aas <- row.srt
    else aas <- sort(unique(x[,rows]))
    if ( !missing(col.srt) ) acod <- col.srt
    else acod <- sort(unique(x[,cols]))

    ## remove rows with missing values
    if ( na.rm ) {
        ina <- is.na(x[,value])
        if ( sum(ina)>0 ) {
            wrn <- paste0(sum(ina), " rows with NA values removed")
            cat(paste("WARNING: ", wrn, "\n"))
            warning(wrn)
            x <- x[!ina,]
        }
    }
    
    ## filter for available classes
    if ( filter ) {
        aas <- aas[aas%in%x[,rows]]
        acod <- acod[acod%in%x[,cols]]
    }
    
    ## result structures
    tt <- matrix(0, ncol=length(acod), nrow=length(aas))
    colnames(tt) <- acod
    rownames(tt) <- aas
    tc <- tp <- tm <- td <- uq <- tt
    tp[] <- 1
    tm[] <- td[] <- NA

    allvals <- list()
    for ( i in seq_along(aas) ) {

        aa <- aas[i]
        allvals[[aa]] <- list()

        for ( j in seq_along(acod) ) {

            cod <- acod[j]

            ## get all rows for the current intersection
            idx <- x[,cols]==cod & x[,rows]==aa
            
            if ( any(is.na(idx)) ) {
                wrn <- paste0(sum(is.na(idx)), " NA in ", aa, "/", cod)
                cat(paste0("WARNING: ", wrn,"\n"))
                warning(wrn)
                idx[is.na(idx)] <- FALSE
            }

            ## local background: column type!
            bgidx <- !idx
            if ( bg ) {
                if ( bg.dir=="col" )
                    bgidx <- x[,cols]==cod
                else if ( bg.dir=="row" )
                    bgidx <- x[,rows]==aa
            }
            if ( !replace ) ## do not include foreground in background!
                bgidx <- bgidx & !idx

            if ( any(is.na(bgidx)) ) {
                wrn <- paste0(sum(is.na(bgidx)), " NA in ",
                              aa, "/", cod, "background")
                cat(paste0("WARNING: ", wrn,"\n"))
                warning(wrn)
                idx[is.na(idx)] <- FALSE
            }

            ## collect unique counter by ID column
            if ( !missing(id) )                 
                uq[i,j] <- length(unique(x[idx,id]))
            
            if (verb>0) cat(paste("testing",cod,"to", aa, "\n"))

            ## get all RAAS values
            y <- unlist(x[  idx, value])
            X <- unlist(x[bgidx, value])

            ## store all values for this class
            allvals[[aa]][[cod]] <- y

            ## value count
            tc[i,j] <- length(y)

            if ( length(y)==0 ) {
                if ( verb>0 )
                    cat(paste0("WARNING: no values for ",aa, "/", cod,"\n"))
                next
            }

            ## MEAN AND MEDIAN
            ## delog for mean and median
            ly <- y
            if ( delog ) 
                ly <- 10^y
            ymn <- mean(ly)
            ymd <- median(ly)
            if ( delog ) {
                ymn <- log10(ymn)
                ymd <- log10(ymd)
            }
            tm[i,j] <- ymn
            td[i,j] <- ymd

            ## T-TEST or RANK SUM TEST
            if ( sum(!is.na(y)) >= min.obs ) {
                
                tts <- use.test(y, X)

                tt[i,j] <- tts$statistic
                tp[i,j] <- tts$p.value
                
                if ( do.plots ) {
                    outfile <- paste0(fname,cod,"_",aa,".png")
                    hcol <- ifelse(tts$statistic<0, "#0000ff","#ff0000")
                    if ( verb>0 ) cat(paste("plotting", outfile, "\n"))
                    png(outfile, width=3, height=2,
                        units="in", res=100)
                    par(mai=c(.5,.15,.25,.15), tcl=-.25, mgp=c(1.3,.3,0))
                    brks <- seq(min(X), max(X), length.out=20)
                    hist(y, freq=FALSE, border=NA,
                         col=paste0(hcol,77),
                         xlim=range(X), main=NA,
                         xlab=xlab, breaks=brks, axes=FALSE)
                    par(new=TRUE)
                    hist(X, freq=FALSE, col="#77777777", border=NA,
                         xlim=range(X), main=NA, xlab=NA, ylab=NA, axes=FALSE,
                         breaks=brks)
                    abline(v=median(X), col=1, lwd=2)
                    abline(v=td[i,j], col=hcol, lwd=2)
                    axis(1)
                    mtext(as.expression(bquote(.(cod) %->% .(aa))),
                          3,-.1, font=2, cex=1.5)
                    legend(ifelse(tts$statistic<0, "topright","topleft"),
                           c(paste0("n=", length(y)),
                             paste0("u/t=", round(tts$statistic,1)),
                             paste0("p=", signif(tts$p.value,1))),
                           bty="n", seg.len=0, x.intersp=0)
                    dev.off()
                }
            }
        }
    }
    if ( verb>0 ) cat(paste("DONE\n"))
    
    ## construct overlap object
    ova <- list()
    ova$p.value <- tp
    ova$statistic <- tt
    ova$count <- tc
    ova$mean <- tm
    ova$median <- td
    ova$unique <- uq

    ova$delog <- delog

    ## sign of change, used in plotOverlaps
    sg <- sign(tt)
    sg[is.na(sg)] <- 1
    sg[sg==0] <- 1
    ova$sign <- sg
    
     ## add counts
    ova$num.target <-
        t(as.matrix(table(x[,cols])[colnames(tp)]))
    ova$num.query <-
        as.matrix(table(x[,rows]))[rownames(tp),,drop=FALSE]


    ## attach all raw values
    ova$ALL <- allvals

    ## multiple testing
    if ( p.adjust=="q" )
        ova$p.value[] <- qvalue::qvalue(c(ova$p.value))$qvalues
    else
        ova$p.value[] <- p.adjust(c(ova$p.value), method=p.adjust)
    ova$p.adjust <- p.adjust
    
    class(ova) <- "clusterOverlaps"
    ova
}

## calculated two-sided hypergeo tests for an amino acid
## table, for each column
#' @param x a matrix of character values, e.g. amino acids, where
#' rows are protein sequences and columns relative positions
#' @param abc list of characters to consider, e.g. all amino acids
aaProfile <- function(x, abc, k=1, p.min, p.adjust="none", verb=0) {

    if ( missing(abc) )
        abc <- sort(unique(c(x)))
    if ( is.null(colnames(x)) )
        colnames(x) <- 1:ncol(x)

    aam <- matrix(1, ncol=ncol(x), nrow=length(abc))
    rownames(aam) <- abc
    colnames(aam) <- colnames(x)
    aac <- aap <- aar <- aam
    aac[] <- 0
    
    for ( i in 1:nrow(aam)   ) {
        if ( verb>0 ) cat(paste("testing", rownames(aam)[i], "\n"))
        for ( j in  1:ncol(aam) ) {

            if ( verb>1 ) cat(paste("\t", colnames(aam)[j], "\n"))
            
            aa <- rownames(aam)[i]
            pos <- colnames(aam)[j]
            
            m <- sum(c(x)==aa) # white balls
            n <- sum(x%in%abc)-m # black balls
            N <- m+n # total number of balls

            q <- sum(x[,j]==aa)    # white balls drawn - AT CURRENT POSITION
            k <- sum(x[,j]%in%abc) # number of balls drawn
            
            aac[i,j] <- q
            ## enriched
            aam[i,j] <- phyper(q=q-1, m=m, n=n, k=k, lower.tail=FALSE)
            ## deprived
            aap[i,j] <- phyper(q=q, m=m, n=n, k=k, lower.tail=TRUE)
            ## frequency ratio
            aar[i,j] <- q/k * N/m
        }
    }
    ## FILTER SIGNIFICANT
    if ( !missing(p.min) ) {
        pval <- aam
        pval[aap<aam] <- aap[aap<aam]
        sig <- apply(pval, 1, function(x) any(x<p.min))
        aam <- aam[sig,,drop=FALSE]
        aap <- aap[sig,,drop=FALSE]
        aac <- aac[sig,,drop=FALSE]
        aar <- aar[sig,,drop=FALSE]
    }

    ## construct overlap class
    ovl <- list()
    ovl$count <- aac
    ovl$ratio <- aar
    ovl$p.value <- aam
    ovl$p.value[aap<aam] <- aap[aap<aam]
    ovl$sign <- ifelse(aap<aam,-1,1)
    ovl$num.query <-
        t(t(table(c(x))[rownames(aam)]))
    ovl$num.target <-
        t(apply(x, 2, function(x) sum(x%in%rownames(aam))))
    ## convert table to matrix
    ovl$num.query <- unclass(ovl$num.query)
    ovl$num.target <- unclass(ovl$num.target)


    ## multiple testing correction
    if ( p.adjust=="q" )
        ovl$p.value[] <- qvalue::qvalue(c(ovl$p.value))$qvalues
    else
        ovl$p.value[] <- p.adjust(c(ovl$p.value), method=p.adjust)
    ovl$p.adjust <- p.adjust

    class(ovl) <- "clusterOverlaps"
    
    ovl
}


## derived from segmenTools::sortOverlaps but trying
## to better handle two-sided tests -
## TODO:
## * integrate in segmenTools
## * allow to only sort enriched or deprived.
sortOverlaps2 <- function (ovl, axis = 2, p.min = 0.05, cut = FALSE,
                           srt, symmetric = "no") {
    
    ## handle triangle matrix
    ## NOTE: currently only produced by segmentOverlaps, where
    ## p.values=1 and counts=0 in the lower triangle
    if (symmetric != "no") {
        if (symmetric == "upper") 
            symm.tri <- lower.tri
        else if (symmetric == "lower") 
            symm.tri <- upper.tri

        ## copy upper to lower
        pvl <- abs(ovl$p.value)
        n <- nrow(pvl)
        m <- ncol(pvl)
        if (n != m) 
            stop("symmetric handling requested for non-symmetric matrix")
        for (i in 1:length(ovl)) {
            x <- ovl[[i]]
            if (inherits(x, "matrix")) 
                if (nrow(x) == n & ncol(x) == m) 
                  x[symm.tri(x)] <- t(x)[symm.tri(x)]
            ovl[[i]] <- x
        }
    }

    ## transpose all, if sorting of x-axis (1) is requested
    if (axis == 1) 
        ovl <- t.clusterOverlaps(ovl)
    
    pvl <- ovl$p.value * -ovl$sign ## NOTE: *sign is NEW FOR TWO-SIDED

    ## sort by significance
    if (missing(srt)) {

        cls.srt <- colnames(pvl)

        if ( is.null(colnames(pvl)) )
            stop("missing column names in p-value matrix")
        
        sig.srt <- NULL
        ## first, get highly significant
        for (cl in cls.srt) {
            tmp.srt <- order(pvl[, cl], decreasing = FALSE)
            ## cut by p value
            sig.srt <- c(sig.srt,
                         tmp.srt[tmp.srt %in% which(abs(pvl[,cl]) < p.min)])
            ## NOTE: abs(pvl) is NEW FOR TWO-SIDED
        }
        ## second, sort rest by increasing pval
        rest.srt <- which(!(1:nrow(pvl)) %in% sig.srt)
        rest.srt <- rest.srt[order(apply(pvl[rest.srt, , drop = FALSE], 
                                         1, max), decreasing = FALSE)]
        new.srt <- sig.srt[!duplicated(sig.srt)]
        if (!cut) 
            new.srt <- c(new.srt, rest.srt)

        ## remember row split between sig and non-sig
        nsig <- sum(!duplicated(sig.srt))
    }
    else {
        ## used passed sorting!
        new.srt <- srt
        nsig <- NULL
        ## 202307 - tested well in clusterGo.R
        ## warning("custom sorting via `srt` is untested!")
    }
    
    ## resort all matrices in overlap structure (overlap, pvalue, jaccard, ...)
    ## TODO: do this safer, check if everything got sorted?
    n <- nrow(pvl)
    m <- ncol(pvl)
    for (i in 1:length(ovl))
        if (inherits(ovl[[i]], "matrix")) { ## check if matrix is of same dim
            if (nrow(ovl[[i]]) == n) 
                ovl[[i]] <- ovl[[i]][new.srt, , drop = FALSE]
            if (symmetric != "no" & ncol(ovl[[i]]) == m) ## symmetric case!
                ovl[[i]] <- ovl[[i]][, new.srt, drop = FALSE]
        }

    ## transpose back
    if (axis == 1) 
        ovl <- t.clusterOverlaps(ovl)

    ## symmetric case: set other to NA
    if (symmetric != "no") {

        ## copy upper to lower
        pvl <- abs(ovl$p.value)
        n <- nrow(pvl)
        m <- ncol(pvl)
        if (n != m) 
            stop("symmetric handling requested for non-symmetric matrix")
        for (i in 1:length(ovl)) {
            x <- ovl[[i]]
            replace <- ifelse(names(ovl)[i] == "p.value", 1, 
                0)
            if (inherits(x, "matrix")) 
                if (nrow(x) == n & ncol(x) == m) 
                  x[symm.tri(x)] <- replace
            ovl[[i]] <- x
        }
    }

    ## add number of sorted sig
     ovl$nsig <- nsig
    ovl$nsigdir <- axis # remember direction
    
    ovl
}


## TODO: use as axis in codon/AA plots?
## plot codons as seqlogo and AA
plotCodonAxis <- function(CODP, aa.cols=aa.cols, cex=1) {

    library(DiffLogo)
    source("~/work/mistrans/scripts/saap_utils.R")
    
    N <- length(CODP)
    layout(cbind(1:N, N +1:N), widths=c(3,1))
    par(mai=rep(0,4), family="monospace")#, yaxs="i")
    for ( i in 1:length(CODP) ) 
        mySeqLogo(CODP[[i]], stackHeight = sumProbabilities, sparse=TRUE,
                  drawLines=1, ylim=c(-.1,1.1))
    for ( i in 1:length(CODP) ) { 
        plot(0,0, axes=FALSE, col=NA)
        shadowtext(0,0,names(CODP)[i], col=aa.cols[names(CODP)[i]], cex=cex,font=2)
    }
}

## generate frequencies at each position
getPFM <- function(aa, alphabet=ASN$chars) {
    ## column-wise table of AA
    ctl <- apply(aa, 2, table)
    ## aaids
    if ( inherits(ctl,"list") ) {
        aaids <- unique(unlist(lapply(ctl, names)))
        ctl <- do.call(cbind, lapply(ctl, function(x) x[aaids]))
        rownames(ctl) <- aaids
    }
    ctl[is.na(ctl)] <- 0

    ## remove all not in alphabet before taking col sum
    ctm <- matrix(0, nrow=length(alphabet), ncol=ncol(aa))
    rownames(ctm) <- alphabet
    ctm[alphabet[alphabet%in%rownames(ctl)],] <-
        ctl[alphabet[alphabet%in%rownames(ctl)],]
    ctm[is.na(ctm)] <- 0

    ## frequencies
    ctf <- t(t(ctm)/apply(ctm,2,sum))
    as.data.frame(ctf)
}

## add significance indicators
diffLogo_addPvals <- function(dfop, ymin, levels=10^-c(3,5,10)) {
    leftOffset = 0
    if (!is.null(dfop$unaligned_from_left)) {
        leftOffset = dfop$unaligned_from_left
    }
    if (!is.null(dfop$unaligned_from_right)) {
        rightOffset = dfop$unaligned_from_right
    }
    npos = ncol(dfop$pwm1)
    for (j in (leftOffset + 1):(npos - rightOffset)) {
        if (dfop$pvals[j] < levels[3]) {
            text(j, ymin, "***", xpd=TRUE, cex=1.5)
        } else  if (dfop$pvals[j] < levels[2]) {
            text(j, ymin, "**", xpd=TRUE, cex=1.5)
        } else  if (dfop$pvals[j] < levels[1]) {
            text(j, ymin, "*", xpd=TRUE, cex=1.5)
        }
    }
}

## plot offsets for diffLogos
DiffLogoObject_getOffsets <- function(diffLogoObj) {
    
    leftOffset = rightOffset = 0
    if (!is.null(diffLogoObj$unaligned_from_left)) {
        leftOffset = diffLogoObj$unaligned_from_left
    }
    if (!is.null(diffLogoObj$unaligned_from_right)) {
        rightOffset = diffLogoObj$unaligned_from_right
    }
    c(leftOffset, rightOffset)
}

myDiffLogo <- function (diffLogoObj, ymin, ymax, sparse = FALSE,
                        diffLogoConfiguration = list()) 
{
    if (!is(diffLogoObj, "DiffLogo")) {
        msg = paste("Expected DiffLogo, but got ", class(diffLogoObj), 
            ". Use #createDiffLogoObject to get an DiffLogo from two PWMs.", 
            sep = "")
        stop(msg)
    }
    if ( missing(ymin) ) { ## NOTE: ylims flipped wrt original function
        ymin = diffLogoObj$ylim.negMax
    }
    if ( missing(ymax) ) {
        ymax = diffLogoObj$ylim.posMax
    }
    ylab = diffLogoObj$ylab
    if (sparse) {
        plot(NA, xlim = c(0.5, diffLogoObj$npos + 0.5), ylim = c(ymin, 
            ymax), xaxt = "n", ylab = "", mgp = c(0, 0.35, 0), 
            tck = -0.02, cex.axis = 0.8, frame.plot = FALSE, 
            xlab = "",
            axes=FALSE) ## NOTE: added axes FALSE to suppress axis settings
        ##axis(2)
    }
    else {
        plot(NA, xlim = c(0.5, diffLogoObj$npos + 0.5), ylim = c(ymin, 
            ymax), xaxt = "n", ylab = ylab, xlab = "Position", 
            frame.plot = FALSE, )
    }
    if (sparse) {
        axis(1, labels = c(rep("", diffLogoObj$npos)),
             at = (1:diffLogoObj$npos), 
             tck = -0.02)
        axis(1, labels = c("", ""), at = c(0, (diffLogoObj$npos + 
            1)), tck = -0)
    }
    else {
        axis(1, labels = c(1:diffLogoObj$npos), at = (1:diffLogoObj$npos))
        axis(1, labels = c("", ""), at = c(0, (diffLogoObj$npos + 
            1)), tck = -0)
    }
    polygon(diffLogoObj$letters, col = diffLogoObj$letters$col, 
        border = FALSE)
    if (!is.null(diffLogoObj$unaligned_from_left) &&
        diffLogoObj$unaligned_from_left > 
        0) {
        rect(0.5, -ymin, diffLogoObj$unaligned_from_left + 0.5, 
            -ymax, col = "gray", border = "gray")
    }
    if (!is.null(diffLogoObj$unaligned_from_right) &&
        diffLogoObj$unaligned_from_right > 0) {
        rect(diffLogoObj$npos - diffLogoObj$unaligned_from_right + 
            0.5, -ymin, diffLogoObj$npos + 0.5, -ymax, col = "gray", 
            border = "gray")
    }
    if (!is.null(diffLogoObj$pvals)) {
        leftOffset = 0
        if (!is.null(diffLogoObj$unaligned_from_left)) {
            leftOffset = diffLogoObj$unaligned_from_left
        }
        if (!is.null(diffLogoObj$unaligned_from_right)) {
            rightOffset = diffLogoObj$unaligned_from_right
        }
        npos = ncol(diffLogoObj$pwm1)
        for (j in (leftOffset + 1):(npos - rightOffset)) {
            if (diffLogoObj$pvals[j] < 0.05) {
                text(j, ymin, "*")
            }
        }
    }
    lines(c(0, diffLogoObj$npos), c(0, 0))
}

## NOTE: customizing DiffLogo::seqLogo
mySeqLogo <- 
function (pwm, sparse = FALSE, drawLines = 0.5, stackHeight = sumProbabilities,
    baseDistribution = probabilities, alphabet = DNA, main = NULL, ylim) 
{

    source("/home/raim/programs/DiffLogo/R/seqLogo.R")
    source("/home/raim/programs/DiffLogo/R/alphabet.R")

    require(DiffLogo)
    # appends the letter which to the object letters
    addLetter = function (letters, letterPolygon, x.pos, y.pos, ht, wt, col="black") {
        x = x.pos + wt * letterPolygon$x
        y = y.pos + ht * letterPolygon$y
        polygons = sum(is.na(x))+1  # a letter can consist of more then one polygon
        letters$x = c(letters$x, NA, x)
        letters$y = c(letters$y, NA, y)
        letters$col = c(letters$col, rep(col, polygons))
        letters
    }
    preconditionPWM = function(pwm) {
        if (any(abs(1 - apply(pwm, 2, sum)) > 0.01)) { # check if the sum of each columns is near 1
            stop("Columns of PWM must add up to 1.0")
        }
    }
    preconditionTransformPWM = function(pwm, alphabet) {
    if (is(pwm, "pwm")) {
        return (pwm@pwm)
    } else if (is(pwm, "data.frame")) {
        return (as.matrix(pwm))
    } else if (is(pwm, "matrix")) {
        print("pwm must be of class matrix or data.frame. Trying to convert")
        return (matrix(pwm,alphabet$size,length(pwm)/alphabet$size))
    }
    return(pwm)
}

    pwm = preconditionTransformPWM(pwm, alphabet)
    preconditionPWM(pwm)
    letters = list(x = NULL, y = NULL, id = NULL, fill = NULL)
    npos = ncol(pwm)
    ylim.negMax = 0
    ylim.posMax = 0
    wt = 1
    x.pos = 0.5
    eps = 0
    heights = c()
    ymins = c()
    ymaxs = c()
    for (j in 1:npos) {
        column = pwm[, j]
        sh = stackHeight(column)
        hts = baseDistribution(column) * sh$height
        letterOrder = order(abs(hts))
        yneg.pos = 0
        ypos.pos = 0
        for (i in 1:alphabet$size) {
            ht = hts[letterOrder[i]]
            y.pos = ypos.pos
            ht = ht - eps
            ypos.pos = ypos.pos + ht + eps
            char = alphabet$chars[letterOrder[i]]
            col = alphabet$cols[letterOrder[i]]
            letters = addLetter(letters, letterPolygons[[char]], 
                x.pos, y.pos, ht, wt * 0.99, col = col)
        }
        x.pos = x.pos + wt
    }
    if ( missing(ylim) ) ylim <- c(0, log2(alphabet$size))
    if (sparse) {
        plot(NA, xlim = c(0.5, x.pos), ylim = ylim, 
            xaxt = "n", ylab = "", mgp = c(0, 0.35, 0), tck = -0.02, 
            cex.axis = 0.8, frame.plot = FALSE, xlab = "", main = main)
    }
    else {
        plot(NA, xlim = c(0.5, x.pos), ylim = ylim, 
            xaxt = "n", ylab = sh$ylab, frame.plot = FALSE, xlab = "Position", 
            main = main)
    }
    for (y in seq(0, log2(alphabet$size), drawLines)) {
        abline(y, 0, col = "gray")
    }
    if (sparse) {
        axis(1, labels = c("", rep("", npos), ""), at = c(0, 
            1:npos, npos + 1), tck = -0.02)
    }
    else {
        axis(1, labels = c("", 1:npos, ""), at = c(0, 1:npos, 
            npos + 1))
    }
    polygon(letters, col = letters$col, border = NA)
}

## for each protein detect and tag windows
get.windows <- function(x, maxdst) {

   ## x=wsite[[pid]]

    if ( nrow(x)==1 ) {
        x$window <- 1
    } else {
        
        ## window ends: larger than max distance to next!
        ends <- which(diff(x$pos)>maxdst)
        
        x$window <- rep(NA, nrow(x))
        x$window[ends] <- 1:length(ends)
        
        ## propagate window number
        win <- 1
        for ( i in 1:nrow(x) ) {
            if ( !is.na(x$window[i]) )
                win <- x$window[i]+1
            else x$window[i] <- win
        }
    }
    x
}
